### DISCUSSION TITLE
blog/top-10-errors-in-r/

### DISCUSSION THREAD

**Discussion by giscus** [AntoineSoetewey/statsandr#104] [General]
# blog/top-10-errors-in-r/

In this post, I highlight the 10 most common errors in R and how to fix them. I also mention a couple of warnings (which are different than errors).

https://statsandr.com/blog/top-10-errors-in-r/

**Comment by ecardenes** [comment]
I think here you should write getwd()
"Alternatively, you can move the dataset in the folder where R is currently working (this can be found with setwd()). See more details on importing a file into R and about the working directory." You can find this text in Problem 9.
Thanks, Antoine!

Elías Cárdenes

**Reply by AntoineSoetewey** [reply]
Dear Elías,

You're completely right, I was too quick on this one.

It has been corrected, thanks for pointing it out!

Regards,
Antoine

**Reply by AntoineSoetewey** [reply]
Any questions @sbkv1?

**Comment by bholden7** [comment]
I've been trying to install cmdstanr and cmdstan on my Mac (Ventura 13.3.1) and get this error message:

Downloading GitHub repo stan-dev/cmdstanr@HEAD
── R CMD build ──────────────────────────────────────────────────────────────────────────
✔  checking for file ‘/private/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T/RtmpADuGtT/remotes4732070f151/stan-dev-cmdstanr-ab30e07/DESCRIPTION’ ...
─  preparing ‘cmdstanr’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Omitted ‘LazyData’ from DESCRIPTION
─  building ‘cmdstanr_0.5.3.tar.gz’
   
Error: 3:49: unexpected string constant
3: if (Sys.getenv("RSTUDIO") == "1" && RSTUDIO_TERM")) &&
4:     Sys.info()["
                                                   ^
Execution halted
Warning: installation of package ‘/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T//RtmpADuGtT/file47356f30d7a/cmdstanr_0.5.3.tar.gz’ had non-zero exit status

I have the latest versions of EVERYTHING, C++ compiler is working fine, etc. Any ideas about a fix?

Many thanks.

**Reply by AntoineSoetewey** [reply]
Hello,

Can you please send me the code you are trying to run?

Regards,
Antoine

**Comment by bholden7** [comment]
Sure, thanks for asking.

I tried this:

```
# we recommend running this is a fresh R session or restarting your current session
install.packages("cmdstanr", repos = c("https://mc-stan.org/r-packages/", getOption("repos")))
```

And got this in response:

```
Warning in install.packages :
  unable to access index for repository https://mc-stan.org/r-packages/bin/macosx/big-sur-arm64/contrib/4.3:
  cannot open URL 'https://mc-stan.org/r-packages/bin/macosx/big-sur-arm64/contrib/4.3/PACKAGES'
installing the source package ‘cmdstanr’

trying URL 'https://mc-stan.org/r-packages/src/contrib/cmdstanr_0.5.3.tar.gz'
Content type 'application/gzip' length 667799 bytes (652 KB)
==================================================
downloaded 652 KB

Error: 3:49: unexpected string constant
3: if (Sys.getenv("RSTUDIO") == "1" && RSTUDIO_TERM")) &&
4:     Sys.info()["
                                                   ^
Execution halted
Warning in install.packages :
  installation of package ‘cmdstanr’ had non-zero exit status

The downloaded source packages are in
	‘/private/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T/RtmpADuGtT/downloaded_packages’
```

Then I tried this:

```
#install.packages("remotes")
remotes::install_github("stan-dev/cmdstanr")
```

And got this response:

```
Downloading GitHub repo ***@***.***
── R CMD build ──────────────────────────────────────────────────────────────────────────
✔  checking for file ‘/private/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T/RtmpADuGtT/remotes473607922cd/stan-dev-cmdstanr-ab30e07/DESCRIPTION’ (566ms)
─  preparing ‘cmdstanr’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Omitted ‘LazyData’ from DESCRIPTION
─  building ‘cmdstanr_0.5.3.tar.gz’
   
Error: 3:49: unexpected string constant
3: if (Sys.getenv("RSTUDIO") == "1" && RSTUDIO_TERM")) &&
4:     Sys.info()["
                                                   ^
Execution halted
Warning: installation of package ‘/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T//RtmpADuGtT/file47330dbf236/cmdstanr_0.5.3.tar.gz’ had non-zero exit status
```

My goal is to install the ‘rethinking’ package, which apparently requires both cmdstan and cmdstanr. I have ‘rstan’ installed and have used it often without problem.

I have the latest versions of R, RStudio, running a MacBook Pro with Ventura 13.3.1. 

I understand that an unexpected string constant error may be a coding error (eg, forgotten parenthesis or quote) but I don’t know where that code is or even if that is the “real” problem. I opened up the security on my Mac as well. I’ve been using RStudio for years with little or no problem (doing mostly data viz and Bayesian analysis) so this is really bothering me! If you can help I will be forever in your debt!

Billy Holden, PhD

**Reply by AntoineSoetewey** [reply]
Dear Billy,

I don't have any experience with the rethinking package.

However, from the [repository of the rethinking package](https://github.com/rmcelreath/rethinking), it seems that the package can be installed with the following code:

```
install.packages(c("coda","mvtnorm","devtools","loo","dagitty","shape"))
devtools::install_github("rmcelreath/rethinking")
```

Have you tried this code?

On the repository, it is also written the following:

> If there are any problems, they likely arise when trying to install rstan, so the rethinking package has little to do with it. See the manual linked above for some hints about getting rstan installed. But always consult the RStan section of the website at mc-stan.org for the latest information on RStan.

Did you follow it this recommendation?

Regards,
Antoine

**Comment by bholden7** [comment]
Hi, and thanks again.

I don’t think the problem is the rethinking package (or rstan, for that matter, which I use often).
I could easily download the other packages (coda et al) but I get the same message when I try to install rethinking, ie, 

Downloading GitHub repo ***@***.***
Skipping 1 packages not available: cmdstanr
── R CMD build ──────────────────────────────────────────────────────────────────────────
✔  checking for file ‘/private/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T/RtmpqDSuZk/remotes46542042d7b4/rmcelreath-rethinking-2f01a9c/DESCRIPTION’ ...
─  preparing ‘rethinking’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘rethinking_2.31.tar.gz’
   
Error: 3:49: unexpected string constant
3: if (Sys.getenv("RSTUDIO") == "1" && RSTUDIO_TERM")) &&
4:     Sys.info()["
                                                   ^
Execution halted
Warning: installation of package ‘/var/folders/66/6yy31gvd2ggdrcmn0v72j0z40000gn/T//RtmpqDSuZk/file465445f7af8/rethinking_2.31.tar.gz’ had non-zero exit status

You can see from the second line I still cannot install cmdstanr (or cmdstan) due to the unexpected string constant error. I believe this is either an R issue or my Mac has an issue, not sure which and not sure how to fix this.

Any thoughts much appreciated.

Thanks.


Billy Holden, PhD

**Reply by AntoineSoetewey** [reply]
Dear Billy,

I executed the following code:

```
remotes::install_github("stan-dev/cmdstanr")
devtools::install_github("rmcelreath/rethinking")

library(rethinking)

f <- alist(
  y ~ dnorm( mu , sigma ),
  mu ~ dnorm( 0 , 10 ),
  sigma ~ dexp( 1 )
)

fit <- quap( 
  f , 
  data=list(y=c(-1,1)) , 
  start=list(mu=0,sigma=1)
)
```

And everything worked smoothly since I can display the results of `fit`:

```
# Quadratic approximate posterior distribution
# 
# Formula:
#   y ~ dnorm(mu, sigma)
# mu ~ dnorm(0, 10)
# sigma ~ dexp(1)
# 
# Posterior means:
#   mu     sigma 
# 0.0000000 0.8392882 
# 
# Log-likelihood: -2.91 
```

So if you have an error with the exact same code, it must come from your settings (your mac, your R, or something else but not from the package I believe).

I don't have any experience with these packages. I suppose you couldn't find a solution on Google, so the only thing I can suggest is to check the issues on the GitHub repositories of the packages (the closed ones too) in case someone else experienced the same issue.

Hope this helps.

Regards,
Antoine

**Comment by bholden7** [comment]
Thanks, Antoine, very helpful. I agree, probably an issue on my Mac (never had a problem with R, but can never discount anything…).
Best,


Billy Holden, PhD

**Comment by ecardenes** [comment]
There is probably a typo in section 1:


"## correct:
install.packages("ggplot2")"

"## also correct:
install.packages("ggplot2")"

My guess is you meant to exemplify single and double quotes ...

Thanks for your work!

Elías Cárdenes

**Reply by AntoineSoetewey** [reply]
Good catch, thanks @ecardenes!

It is corrected.
